qiskit_experiments/stored_data/stored_data.py

```diff
@@ -0,0 +1,1047 @@
+                    "service is not changed if one already exists.",
+                    data.backend(),
+                    self.backend,
+                )
+            self._backend = data.backend()
+            if not self._service:
+                self._set_service_from_backend(self._backend)
+
+            self._jobs[data.job_id()] = data
+            self._job_futures.append(
+                (
+                    data,
+                    self._executor.submit(
+                        self._wait_for_job, data, post_processing_callback, **kwargs
+                    ),
+                )
+            )
+            if self.auto_save:
+                self.save()
+            return
+
+        if isinstance(data, dict):
+            self._add_single_data(data)
+        elif isinstance(data, Result):
+            self._add_result_data(data)
+        elif isinstance(data, list):
+            # TODO use loop instead of recursion for fewer save()
+            for dat in data:
+                self.add_data(dat)
+        else:
+            raise TypeError(f"Invalid data type {type(data)}.")
+
+        if post_processing_callback is not None:
+            post_processing_callback(self, **kwargs)
+
+    def _wait_for_job(
+        self,
+        job: Union[Job, BaseJob],
+        job_done_callback: Optional[Callable] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Wait for a job to finish.
+
+        Args:
+            job: Job to wait for.
+            job_done_callback: Callback function to invoke when job finishes.
+            **kwargs: Keyword arguments to be passed to the callback function.
+
+        Raises:
+            Exception: If post processing failed.
+        """
+        LOG.debug("Waiting for job %s to finish.", job.job_id())
+        try:
+            job_result = job.result()
+            with self._data.lock:
+                # Hold the lock so we add the block of results together.
+                self._add_result_data(job_result)
+        except JobError as err:
+            LOG.warning("Job %s failed: %s", job.job_id(), str(err))
+            return
+
+        try:
+            if job_done_callback:
+                job_done_callback(self, **kwargs)
+        except Exception:  # pylint: disable=broad-except
+            LOG.warning("Post processing function failed:\n%s", traceback.format_exc())
+            raise
+
+    def _add_result_data(self, result: Result) -> None:
+        """Add data from a Result object
+
+        Args:
+            result: Result object containing data to be added.
+        """
+        if result.job_id not in self._jobs:
+            self._jobs[result.job_id] = None
+        for i in range(len(result.results)):
+            data = result.data(i)
+            data["job_id"] = result.job_id
+            if "counts" in data:
+                # Format to Counts object rather than hex dict
+                data["counts"] = result.get_counts(i)
```

Comment:
Yes, in calibration experiment, sometimes we use IQ data because discriminator is not calibrated at the very beginning of experiments. And we might use `memory=True` option to keep measured bitstream for applications such as random number generator.

Reply:
Yes please add the memory. We need level 1 data.

Reply:
Added in e916c5f. 
I remember why there is only `get_counts` - it is to reformat counts, not to add counts. `data` always had `memory` since it's a copy of result data, but I added `get_memory` to get the nicer format.